pkg/download/consistent_hashing.go

```diff
@@ -0,0 +1,263 @@
+package download
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"hash/fnv"
+	"io"
+	"net/http"
+	"net/url"
+	"runtime"
+	"strconv"
+
+	jump "github.com/dgryski/go-jump"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/replicate/pget/pkg/client"
+	"github.com/replicate/pget/pkg/logging"
+)
+
+type ConsistentHashingMode struct {
+	Client *client.HTTPClient
+	Options
+}
+
+func GetConsistentHashingMode(opts Options) (Strategy, error) {
+	if opts.SliceSize == 0 {
+		return nil, fmt.Errorf("Must specify slice size in consistent hashing mode")
+	}
+	if opts.Semaphore != nil && opts.MaxConcurrency == 0 {
+		return nil, fmt.Errorf("If you provide a semaphore you must specify MaxConcurrency")
+	}
+	client := client.NewHTTPClient(opts.Client)
+	return &ConsistentHashingMode{
+		Client:  client,
+		Options: opts,
+	}, nil
+}
+
+func (m *ConsistentHashingMode) maxConcurrency() int {
+	maxChunks := m.MaxConcurrency
+	if maxChunks == 0 {
+		return runtime.NumCPU() * 4
+	}
+	return maxChunks
+}
+
+func (m *ConsistentHashingMode) minChunkSize() int64 {
+	minChunkSize := m.MinChunkSize
+	if minChunkSize == 0 {
+		minChunkSize = defaultMinChunkSize
+	}
+	if minChunkSize > m.SliceSize {
+		minChunkSize = m.SliceSize
+	}
+	return minChunkSize
+}
+
+func (m *ConsistentHashingMode) getFileSizeFromContentRange(contentRange string) (int64, error) {
+	groups := contentRangeRegexp.FindStringSubmatch(contentRange)
+	if groups == nil {
+		return -1, fmt.Errorf("couldn't parse Content-Range: %s", contentRange)
+	}
+	return strconv.ParseInt(groups[1], 10, 64)
+}
+
+func (m *ConsistentHashingMode) Fetch(ctx context.Context, urlString string) (io.Reader, int64, error) {
+	logger := logging.GetLogger()
+
+	parsed, err := url.Parse(urlString)
+	if err != nil {
+		return nil, -1, err
+	}
+	shouldContinue := false
+	for _, host := range m.DomainsToCache {
+		if host == parsed.Host {
+			shouldContinue = true
+			break
+		}
+	}
+	if !shouldContinue {
+		return nil, -1, fmt.Errorf("ConsistentHashingMode not implemented for domains outside DomainsToCache")
+	}
```

Comment:
We could make this work for *anything* in reality with a little thought to the cache nginx configs. I am not sure we want to though.

Reply:
yeah I'd like this to fall back to BufferMode in future